### DISCUSSION TITLE
Zero-Tolerance NSFW - Is it possible?

### DISCUSSION THREAD

**Discussion by ProgramDeez** [lllyasviel/Fooocus#3265] [Q&A] [answered]
Curious if it is possible to 100% ensure no adult content, NSFW, or age restricted themes are outputted during image generation/preview?

I know that Fooocus has built in NSFW detection that auto outputs an all black image, however certain prompting can still result in NSFW content generation (generally subtle artistic nudity). I have tried a few different ways to achieve this including use of various models, loras, and strongly weighted negative prompt terms.

Even with all of that, NSFW can sometimes still be generated (as low as ~1 in every 300-400 images).

I understand that this is primarily prompt/model dependent (how it’s trained), but just wanted to see if anyone has found success within Fooocus or via any particular model, checkpoint combination, embedding, lora, or alternative filters such as (https://github.com/EsteveSegura/NSFW-StableDifussion-Filter.git or https://github.com/thefcraft/nsfw-prompt-detection-sd.git), etc. that is ultra strict and will make sure that absolutely no NSFW content is generated?

Appreciate it in advance!

**Comment by mrhan1993** [comment]
I think you need something like ` Intelligent Auditing`, some cloud company have this

**Comment by DavidDragonsage** [comment]
@ProgramDeez  I don't have a magic solution but I never saw any nudity in Fooocus at all until I started using Juggernaut X. It has a annoying tendency to produce NSFW images - some distorted ones at that - at the slightest provocation.

I did have some luck with reducing Juggernaut X's NSFW enthusiasm using this as a wildcard in the negative prompt:
"(NSFW, naked, nude, nudity:1.8)"
Notice the ridiculously high weighting! Anything less and Juggernaut X would cross the line.

Apart from Juggernaut X, I have produced hundreds of images with Juggernaut 8, Juggernaut 9, FaeTastic, Realistic Vision XL, Realistic Stock Photo, SDXL, Dreamshaper, and now Playground 2.5, and never seen any nudity.

**Comment by mashb1t** [comment, marked as answer]
100% prevention isn't possible, especially when users try to work their ways around with some tricks. What the current implementation of the nsfw filter can prevent is generation of full nudity and genitalia as well as nipples on bare skin, but not all implicit nudity such as bikinis / underwear / similar.

**Reply by ProgramDeez** [reply]
Appreciate the insight. Wanted to follow up by asking if as an alternative it might be possible to initiate the NSFW checker that is built to Fooocus earlier within the image generation process (such as when a refiner switch takes place, or at 50% of step value) instead of after the image is fully formed? 

Seems as if it could be done, but not entirely knowledgeable if the checker would have issues assessing an under-developed image?

**Reply by mashb1t** [reply]
@ProgramDeez keep in mind thst nsfw checks take about 1+ second, which would impact image generation if done too frequently, which is why only the final image is checked. Feel free to adjust the code to your liking though.

**Reply by ProgramDeez** [reply]
Good point. Apologize for the continued consultation here, but was hoping to further clarify how I might be able to exactly implement, as I had been trying to do so prior to reaching out, but am running into some issues.

Was working in the 'async_worker.py' file, and messing with 'yield_result()'. Am I off base on that?

**Reply by mashb1t** [reply]
Yes, this is the correct file.

You need to add 
```
yield_result(async_task, cn_img, black_out_nsfw, do_not_show_finished_images=True)
continue
```

where it suits you in https://github.com/lllyasviel/Fooocus/blob/main/modules/async_worker.py#L868 or in the callback, as this is the code which is being called for each step.

**Reply by ProgramDeez** [reply]
Spent some time working to integrate, however it looks like the code you provided was showing a syntax error due to 'yield_result()' not being a loop. Tried both return and pass, which lets it run however an error ends up being thrown when trying to call 'yield_result()' during image generation: 

```
`*\modules\async_worker.py", line 81, in yield_result
    imgs = default_censor(imgs) . . .
 
*\packages\transformers\image_utils.py", line 132, in make_list_of_images
    raise ValueError(
ValueError: Invalid image type. Expected either PIL.Image.Image, numpy.ndarray, torch.Tensor, tf.Tensor or jax.ndarray, but got <class 'list'>` 
```

From what I am able to gather, it seems as if 'default_censor()' is expecting an image or an array of images but is receiving the class 'list' instead. I was trying to define 'cn_img' appropriately so that it would be valid, but I haven't had any luck. Tried integrating into the the callback function and also directly into the step code execution, but was running into the same issue. 

Curious if I am misunderstanding the error and the fix/implementation is easier than I am making it or if implementing the safety check during image progression requires manipulating the checker itself to allow for this vs just assessing the final image?

**Reply by mashb1t** [reply]
you can only yield one image at a time, not a list of images

**Reply by ProgramDeez** [reply]
Gotcha, but calling 'yield_result()' is yielding the class 'list' instead of yielding an image when it is called from within the callback function. 

Trying to identify exactly what is within this list, but it seems as if for some reason 'cn_img' is not properly being defined/not allowing for the check to progress.
